src/Features/Core/Portable/Snippets/RoslynLSPSnippetConverter.cs

```diff
@@ -0,0 +1,174 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Snippets
+{
+    internal static class RoslynLSPSnippetConverter
+    {
+        public static async Task<string> GenerateLSPSnippetAsync(Document document, int caretPosition, ImmutableArray<SnippetPlaceholder> placeholders, TextChange collapsedTextChange)
+        {
+            var extendedTextChange = await ExtendSnippetTextChangeAsync(document, collapsedTextChange, placeholders, caretPosition).ConfigureAwait(false);
+            return ConvertToLSPSnippetString(extendedTextChange, placeholders, caretPosition);
+        }
+
+        /// <summary>
+        /// Iterates through every index in the snippet string and determines where the
+        /// LSP formatted chunks should be inserted for each placeholder.
+        /// </summary>
+        private static string ConvertToLSPSnippetString(TextChange textChange, ImmutableArray<SnippetPlaceholder> placeholders, int caretPosition)
+        {
+            var textChangeStart = textChange.Span.Start;
+            var textChangeText = textChange.NewText!;
+            using var _ = PooledStringBuilder.GetInstance(out var lspSnippetString);
+            var map = GetMapOfSpanStartsToLSPStringItem(placeholders, textChangeStart);
+
+            for (var i = 0; i < textChangeText.Length + 1;)
+            {
+                if (i == caretPosition - textChangeStart)
+                {
+                    lspSnippetString.Append("$0");
+
+                    // Special case for cursor position since they will occur between positions
+                    // so we still wants to insert the character following the cursor position.
```

Comment:
why is this only for cursor position?  can't this happen for placeholders as well?  doc.